components/script/dom/validitystate.rs

```diff
@@ -69,11 +71,36 @@ impl ValidityState {
 impl ValidityStateMethods for ValidityState {
     // https://html.spec.whatwg.org/multipage/#dom-validitystate-valuemissing
     fn ValueMissing(&self) -> bool {
+        if self.element.has_attribute(&local_name!("required")) {
+            let v = self.element.get_string_attribute(&local_name!("value"));
+            let n = self.element.get_string_attribute(&local_name!("name"));
+            if v.is_empty() {
+                println!("{} Value missing!", n);
+                return true;
+            }
+        }
         false
     }
 
     // https://html.spec.whatwg.org/multipage/#dom-validitystate-typemismatch
     fn TypeMismatch(&self) -> bool {
+        if self.element.has_attribute(&local_name!("type")) {
+            let content_type = self.element.get_string_attribute(&local_name!("type"));
+            let content_value = self.element.get_string_attribute(&local_name!("value"));
+            let re = match content_type.to_string().as_ref() {
+                    "email" => Regex::new(r"^\w+@[a-zA-Z_]+?\.[a-zA-Z]{2,3}$").unwrap(),
+                    "url" => Regex::new(r"^[a-zA-Z0-9\-\.]+\.(com|org|net|mil|edu|COM|ORG|NET|MIL|EDU)$").unwrap(),
+                    "date" => Regex::new(r"^\d{1,2}\/\d{1,2}\/\d{4}$").unwrap(),
+                    "tel" => Regex::new(r"^\d{3}-\d{2}-\d{4}$").unwrap(),
```

Comment:
Unfortunately these regexes need to be either much more complex, or much less strict. I lean towards much less strict - URLs can check if there's a period present; emails can check if there's an @; telephone numbers should not contain letters.